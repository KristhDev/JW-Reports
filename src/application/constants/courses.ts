export const coursesMessages = {
    EXPORTED_SUCCESS: 'Has exportado tus cursos correctamente. El archivo se encuentra en su carpeta de descargas.',
    PERSON_NAME_REQUIRED: 'El nombre de la persona es requerido.',
    PERSON_NAME_MIN_LENGTH: 'El nombre de la persona debe tener al menos 2 caracteres.',
    PERSON_ABOUT_REQUIRED: 'La información de la persona es requerida.',
    PERSON_ABOUT_MIN_LENGTH: 'La información de la persona debe tener al menos 10 caracteres.',
    PERSON_ADDRESS_REQUIRED: 'La dirección es requerida.',
    PERSON_ADDRESS_MIN_LENGTH: 'La dirección debe tener al menos 10 caracteres.',
    PUBLICATION_REQUIRED: 'La publicación de estudio no puede estar vacía.',
    PUBLICATION_MIN_LENGTH: 'La publicación de estudio debe tener al menos 5 caracteres.',
    UNSELECTED: 'No hay un curso seleccionado.',
    UNSELECTED_FINISH_OR_START: 'No pudes terminar o comenzar de nuevo un curso suspendido.',
    UNSELECTED_DELETE: 'No hay un curso seleccionado para eliminar.',
    UNSELECTED_UPDATE: 'No hay un curso seleccionado para actualizar.',
    FINISHED: 'No puedes suspender o renovar un curso terminado.',
    SUSPENDED_SUCCESS: 'Has suspendido el curso correctamente.',
    RENEW_SUCCESS: 'Has renovado el curso correctamente.',
    DELETED_SUCCESS: 'Has eliminado el curso correctamente.',
    FINISHED_SUCCESS: 'Has terminado el curso correctamente.',
    RESTARTED_SUCCESS: 'Has comenzado de nuevo el curso correctamente.',
    ADDED_SUCCESS: 'Has agregado el curso correctamente.',
    UPDATED_SUCCESS: 'Has actualizado el curso correctamente.',
}