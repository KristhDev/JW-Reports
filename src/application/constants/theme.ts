import { ItemOption } from '../../infrasturcture/interfaces';

/**
 * The theme options to select
 */
export const THEME_OPTIONS: ItemOption[] = [
    { label: 'Modo predeterminado', value: 'default' },
    { label: 'Modo claro', value: 'light' },
    { label: 'Modo oscuro', value: 'dark' }
];