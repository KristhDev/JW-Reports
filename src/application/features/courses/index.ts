export {
    default as coursesReducer,
    INIT_COURSE,
    COURSES_INITIAL_STATE,
    addCourse,
    addCourses,
    addLastLessonInCourse,
    clearCourses,
    removeCourse,
    removeCourses,
    replaceLastLessonInCourse,
    setCourseFilter,
    setCourses,
    setCoursesPagination,
    setCoursesScreenHistory,
    setHasMoreCourses,
    setIsCourseDeleting,
    setIsCourseLoading,
    setIsCoursesExporting,
    setIsCoursesLoading,
    setRefreshCourses,
    setSelectedCourse,
    updateCourse,
    updateLastLessonInCourse
} from './slice';

export * from './types';