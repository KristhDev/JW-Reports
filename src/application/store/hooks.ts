import { useDispatch, useSelector } from 'react-redux';

/* Store */
import { AppDispatch, RootState } from './store';

/* Defined types hooks */
export const useAppDispatch = useDispatch.withTypes<AppDispatch>();
export const useAppSelector = useSelector.withTypes<RootState>();