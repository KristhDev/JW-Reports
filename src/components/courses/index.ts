export { CourseCard } from './CourseCard';
export { CoursesList } from './CoursesList';