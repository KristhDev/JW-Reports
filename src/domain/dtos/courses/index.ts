export * from './active-suspend.dto';
export * from './create.dto';
export * from './finish-start.dto';
export * from './update.dto';