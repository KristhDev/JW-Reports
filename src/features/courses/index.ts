export {
    default as coursesReducer,
    INIT_COURSE,
    INIT_LESSON,
    addCourse,
    addCourses,
    addLesson,
    addLessons,
    clearCourses,
    removeCourse,
    removeCourses,
    removeLesson,
    removeLessons,
    setCourseFilter,
    setCourses,
    setCoursesPagination,
    setCoursesScreenHistory,
    setHasMoreCourses,
    setHasMoreLessons,
    setIsCourseDeleting,
    setIsCourseLoading,
    setIsCoursesLoading,
    setIsLessonDeleting,
    setIsLessonLoading,
    setIsLessonsLoading,
    setLessons,
    setLessonsPagination,
    setRefreshCourses,
    setRefreshLessons,
    setSelectedCourse,
    setSelectedLesson,
    updateCourse,
    updateLesson
} from './slice';