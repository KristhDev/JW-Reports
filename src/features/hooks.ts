import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';

/* Store */
import { AppDispatch, RootState } from './store';

/* Defined types hooks */
export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;