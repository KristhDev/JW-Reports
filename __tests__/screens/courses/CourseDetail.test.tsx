import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react-native';
import hexToRgba from 'hex-to-rgba';
import dayjs from 'dayjs';

/* Screens */
import { CourseDetail } from '../../../src/screens/courses';

/* Hooks */
import { useCourses, useTheme } from '../../../src/hooks';

/* Theme */
import { darkColors } from '../../../src/theme';

/* Setup */
import { navigateMock } from '../../../jest.setup';

/* Mocks */
import { courseSelectedStateMock, setSelectedCourseMock, setSelectedLessonMock } from '../../mocks';

/* Mock hooks */
jest.mock('../../../src/hooks/useCourses.ts');
jest.mock('../../../src/hooks/useTheme.ts');

describe('Test in <CourseDetail /> screen', () => {
    (useCourses as jest.Mock).mockReturnValue({
        state: courseSelectedStateMock,
        activeOrSuspendCourse: jest.fn(),
        finishOrStartCourse: jest.fn(),
        setSelectedCourse: setSelectedCourseMock,
        setSelectedLesson: setSelectedLessonMock
    });

    (useTheme as jest.Mock).mockReturnValue({
        state: { colors: darkColors },
        BUTTON_TRANSLUCENT_COLOR: hexToRgba(darkColors.button, 0.50),
        BUTTON_TRANSPARENT_COLOR: 'rgba(255, 255, 255, 0.15)'
    });

    beforeEach(() => {
        render(<CourseDetail />);

        jest.clearAllMocks();
    });

    it('should to match snapshot', () => {
        expect(screen.toJSON()).toMatchSnapshot();
    });

    it('should render course data', () => {
        const selectedCourse = courseSelectedStateMock.selectedCourse;

        const statusCourseText = (selectedCourse.finished)
            ? 'Terminado'
            : (selectedCourse.suspended)
                ? 'Suspendido'
                : 'En curso';

        /* Get elements with data of course */
        const title = screen.getByTestId('title-text');
        const publication = screen.getByTestId('info-text-text');
        const status = screen.getByTestId('course-detail-status');
        const aboutSection = screen.getByTestId('course-detail-about-section');
        const addressSection = screen.getByTestId('course-detail-address-section');
        const date = screen.getByTestId('course-detail-text-date');

        /* Check if title, publication and status are exists and contain respective values */
        expect(title).toBeTruthy();
        expect(title.props.children).toBe(selectedCourse.personName.toUpperCase());
        expect(publication).toBeTruthy();
        expect(publication.props.children).toBe(selectedCourse.publication.toUpperCase());
        expect(status).toBeTruthy();
        expect(status.props.children.join('')).toBe(`Estado del curso: ${ statusCourseText }`);

        /* Check if about section exists and contain respective value */
        expect(aboutSection).toBeTruthy();
        expect(aboutSection.props.children[0].props.children.join('')).toBe(`Información de ${ selectedCourse.personName }:`);
        expect(aboutSection.props.children[1].props.children).toBe(selectedCourse.personAbout);

        /* Check if address section exists and contain respective value */
        expect(addressSection).toBeTruthy();
        expect(addressSection.props.children[0].props.children).toBe('Dirección:');
        expect(addressSection.props.children[1].props.children).toBe(selectedCourse.personAddress);

        /* Check if date exists and contain respective value */
        expect(date).toBeTruthy();
        expect(date.props.children).toBe(`${ dayjs(selectedCourse.createdAt).format('DD/MM/YYYY') }`);
    });

    it('should render respective touchable of flag course.finished', () => {
        const selectedCourse = courseSelectedStateMock.selectedCourse;

        const statusCourseAsk = (selectedCourse.finished)
            ? '¿Comenzar de nuevo?'
            : (selectedCourse.suspended)
                ? '¿Continuar?'
                : '¿Suspender?';

        /* Get touchable */
        const toucable = screen.getByTestId('course-detail-status-touchable');

        /* Check if toucable exists and contain respective ask */
        expect(toucable).toBeTruthy();
        expect(toucable.props.children[0].props.children).toBe(statusCourseAsk);
    });

    it('should render last_lesson data if exists', () => {
        const lastLesson = courseSelectedStateMock.selectedCourse.lastLesson;

        if (!lastLesson) {
            expect(true).toBeTruthy();
            return;
        }

        const statusText =  (lastLesson.done)
            ? 'Clase impartida'
            : `Próxima clase ${ dayjs(lastLesson.nextLesson).format('DD/MM/YYYY') }`

        /* Get elements with data of last lesson */
        const status = screen.getByTestId('course-detail-last-lesson-status');
        const description = screen.getByTestId('course-detail-last-lesson-description');

        /* Check if status and description exists and contain respective values */
        expect(status).toBeTruthy();
        expect(status.props.children).toBe(statusText);
        expect(description).toBeTruthy();
        expect(description.props.children).toBe(lastLesson.description);
    });

    it('should call navigate when lessons link is pressed', () => {

        /* Get touchable */
        const touchable = screen.getByTestId('course-detail-lessons-touchable');
        fireEvent.press(touchable);

        /* Check if navigate is called one time with respective value */
        expect(navigateMock).toHaveBeenCalledTimes(1);
        expect(navigateMock).toHaveBeenCalledWith('LessonsScreen');
    });

    it('should call setSelectedLesson and navigate when add lesson link is pressed', () => {

        /* Get touchable */
        const touchable = screen.getByTestId('course-detail-add-lesson-touchable');
        fireEvent.press(touchable);

        /* Check if selectedLesson is called one time with respective value */
        expect(setSelectedLessonMock).toHaveBeenCalledTimes(1);
        expect(setSelectedLessonMock).toHaveBeenCalledWith({
            ...courseSelectedStateMock.selectedLesson,
            nextLesson: expect.any(String),
            createdAt: expect.any(String),
            updatedAt: expect.any(String)
        });

        /* Check if navigate is called one time with respective value */
        expect(navigateMock).toHaveBeenCalledTimes(1);
        expect(navigateMock).toHaveBeenCalledWith('AddOrEditLessonScreen');
    });
});