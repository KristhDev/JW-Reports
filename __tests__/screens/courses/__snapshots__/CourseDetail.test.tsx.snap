// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <CourseDetail /> screen should to match snapshot 1`] = `
[
  <RCTScrollView
    contentContainerStyle={
      {
        "alignItems": "center",
        "flexGrow": 1,
        "padding": 24,
        "paddingBottom": 100,
      }
    }
    overScrollMode="never"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          [
            {
              "alignItems": "flex-start",
              "marginBottom": 40,
              "width": "100%",
            },
            {
              "paddingBottom": 0,
              "paddingTop": 8,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#FFFFFF",
                "fontSize": 38,
                "fontWeight": "bold",
              },
              {
                "fontSize": 24,
              },
            ]
          }
          testID="title-text"
        >
          FRED CREMIN
        </Text>
      </View>
      <View
        style={
          [
            {
              "padding": 16,
            },
            {
              "padding": 0,
              "paddingBottom": 24,
              "width": "100%",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#BFBFBF",
                "fontSize": 17,
                "textAlign": "center",
              },
              {
                "fontSize": 16,
                "fontWeight": "bold",
                "textAlign": "left",
              },
            ]
          }
          testID="info-text-text"
        >
          ANIMI VOLUPTAS OMNIS
        </Text>
      </View>
      <View
        style={
          {
            "paddingBottom": 24,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "color": "#FFFFFF",
              "fontSize": 19,
              "fontWeight": "bold",
              "marginBottom": 0,
            }
          }
          testID="course-detail-status"
        >
          Estado del curso: 
          En curso
        </Text>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
          testID="course-detail-status-touchable"
        >
          <Text
            style={
              {
                "color": "#A0B9E2",
                "fontSize": 19,
              }
            }
          >
            ¿Suspender?
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "paddingBottom": 24,
            "width": "100%",
          }
        }
        testID="course-detail-about-section"
      >
        <Text
          style={
            {
              "color": "#FFFFFF",
              "fontSize": 19,
              "fontWeight": "bold",
              "marginBottom": 8,
            }
          }
        >
          Información de 
          Fred Cremin
          :
        </Text>
        <Text
          style={
            {
              "color": "#FFFFFF",
              "fontSize": 19,
            }
          }
        >
          Voluptatem dolorem neque quia omnis odit quos non. Sed rerum est est rem et. Ut aut repellat expedita commodi ipsum. In quisquam excepturi rerum debitis. Beatae rem saepe aut molestiae.
        </Text>
      </View>
      <View
        style={
          {
            "paddingBottom": 24,
            "width": "100%",
          }
        }
        testID="course-detail-address-section"
      >
        <Text
          style={
            {
              "color": "#FFFFFF",
              "fontSize": 19,
              "fontWeight": "bold",
              "marginBottom": 8,
            }
          }
        >
          Dirección:
        </Text>
        <Text
          style={
            {
              "color": "#FFFFFF",
              "fontSize": 19,
            }
          }
        >
          Non ut fuga dolores. Laborum veniam quae minus consequuntur quisquam temporibus et. Est accusamus omnis facere facilis labore.
        </Text>
      </View>
      <View
        style={
          {
            "paddingBottom": 24,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "color": "#FFFFFF",
              "fontSize": 19,
              "fontWeight": "bold",
              "marginBottom": 8,
            }
          }
        >
          Última clase:
        </Text>
        <View
          style={
            {
              "borderColor": "#FFFFFF",
              "borderWidth": 0.5,
            }
          }
        >
          <View
            style={
              {
                "backgroundColor": "#292929",
              }
            }
          >
            <Text
              style={
                {
                  "color": "#FFFFFF",
                  "fontSize": 24,
                  "fontWeight": "bold",
                  "padding": 8,
                  "textAlign": "center",
                }
              }
              testID="course-detail-last-lesson-status"
            >
              Próxima clase 31/12/2020
            </Text>
          </View>
          <View>
            <Text
              style={
                {
                  "color": "#FFFFFF",
                  "fontSize": 19,
                  "padding": 24,
                }
              }
              testID="course-detail-last-lesson-description"
            >
              Nemo et et non nesciunt ipsum sed rerum vero. Nostrum quae officiis qui nihil dignissimos consectetur aperiam. Vel harum ex. Ipsum ratione necessitatibus eum sint vero enim et consequatur. Ab autem rerum numquam ut. Laboriosam nihil est exercitationem unde aut reiciendis odit.
            </Text>
          </View>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "marginTop": 24,
              "opacity": 1,
            }
          }
          testID="course-detail-lessons-touchable"
        >
          <Text
            style={
              {
                "color": "#A0B9E2",
                "fontSize": 19,
              }
            }
          >
            Ver todas las clases
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "marginTop": 8,
              "opacity": 1,
            }
          }
          testID="course-detail-add-lesson-touchable"
        >
          <Text
            style={
              {
                "color": "#A0B9E2",
                "fontSize": 19,
              }
            }
          >
            Agregar clase
          </Text>
        </View>
      </View>
      <Text
        style={
          {
            "bottom": 24,
            "color": "#B4B4B4",
            "fontSize": 16,
            "position": "absolute",
            "right": 24,
          }
        }
        testID="course-detail-text-date"
      >
        26/08/2021
      </Text>
    </View>
  </RCTScrollView>,
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    statusBarTranslucent={true}
    transparent={true}
    visible={false}
  />,
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    statusBarTranslucent={true}
    transparent={true}
    visible={false}
  />,
]
`;
